### ISSUE TITLE
Compose Custom Loss with Softmax

### ISSUE THREAD

**Issue by RishabGargeya** [apache/incubator-mxnet#2677]
Hello!

I'm trying to create a custom loss function using symbolic variables. The loss function is the sum of the softmax error and other values computed from internal layers. I know I cannot achieve this with `SoftmaxOutput` so I am using `SoftmaxActivation` as my activation function after the final fc layer and `softmax_cross_entropy` as the function to compute softmax loss stacked with `MakeLoss`. However, I am unable to run this pipeline without getting an error:

`Check failed: (this->shape_.Size()) == (shape.Size()) TBlob.get_with_shape: new and old shape do not match total elements`

I know there is a shape mismatch but I'm not sure what the issue is

Here is some sample code showing the basic structure I want to achieve with the network:

```
data = mx.symbol.Variable(name='data')
fc1 = mx.symbol.FullyConnected(data=data, num_hidden=num_hidden, name='fc1')
act = mx.symbol.Activation(data=fc1, act_type='sigmoid')
fc2 = mx.symbol.FullyConnected(data=act, num_hidden=2, name='fc2')
label = mx.symbol.Variable(name='softmax_label')
softmax_activation = mx.symbol.SoftmaxActivation(data=fc2)
loss_1 = mx.symbol.softmax_cross_entropy(lhs=softmax_activation, rhs=label)
loss_2 = mx.symbol.sum(ms.symbol.square(fc1 - 0.5))
return mx.symbol.MakeLoss(data=loss_1+loss_2, name='makeloss')
```

Does this make sense? I'm pretty sure this is possible, but I'm not sure where I went wrong.

On its own, the following does not compile:

```
data = mx.symbol.Variable(name='data')
fc1 = mx.symbol.FullyConnected(data=act, num_hidden=2, name='fc1')
label = mx.symbol.Variable(name='softmax_label')
softmax_activation = mx.symbol.SoftmaxActivation(data=fc1)
loss_1 = mx.symbol.softmax_cross_entropy(lhs=softmax_activation, rhs=label)
return mx.symbol.MakeLoss(data=loss_1, name='makeloss')
```

while this does:

```
data = mx.symbol.Variable(name='data')
fc1 = mx.symbol.FullyConnected(data=act, num_hidden=2, name='fc1')
label = mx.symbol.Variable(name='softmax_label')
return mx.symbol.SoftmaxOutput(data=fc1, label=label)
```

**Comment by piiswrong** [comment]
You can use softmaxoutput for this. Multiple outputs are summed during backward

**Comment by RishabGargeya** [comment]
Hello,

Thank you for the response! However, I am still a bit confused.

My aimed objective function is : softmax loss + custom value computed on internal fc layer

SoftmaxOutput does not output the raw loss value, so I cannot add this to another custom value and run MakeLoss on that? Or are you saying I don't need to explicitly add these symbols together

**Comment by piiswrong** [comment]
You don't need makeloss for simple addition between objectives. They are implicitly added during backprop.
You just need to compute metric correctly for logging

**Comment by CitronMan** [comment]
@RishabGargeya have you figure out this problem? I have the same question....

**Comment by zihaolucky** [comment]
@piiswrong I've checked some similar issues like **SoftmaxOutput as input** and I have something different that needs your insights.

I think sometimes the first softmax layer and the downstream layer are from different but related tasks, like classifying different kinds of cats and dogs together, in which we might want a cat vs dog binary classifier and then use its output as input feature in the later fine-grained cats/dogs task. 

Using `mx.sym.Group([SoftmaxOutput, SomeotherOutput])` would take these tasks all together but what if we want to implement the idea above? 

```python
data = mx.symbol.Variable(name='data')
label1 = mx.symbol.Variable(name='softmax_label')    # dog or cat, binary indicator.
label2 = mx.symbol.Variable(name='softmax_label2') # dogs and cats

# some common layers.
conv = mx.symbol.Convolution(data=data, num_filter=30, kernel=(5, 5), stride=(2,2))
act = mx.symbol.Activation(data = conv, act_type='relu')
pooling = mx.symbol.Pooling(data=act, kernel=(2, 2), stride=(2, 2), pool_type='max')
conv = mx.symbol.Convolution(data=pooling, num_filter=60, kernel=(3, 3), stride=(1,1), pad=(1, 1))
act = mx.symbol.Activation(data = conv, act_type='relu')
pooling = mx.symbol.Pooling(data=act, global_pool=True, kernel=(2, 2), pool_type='avg')
flat = mx.symbol.Flatten(data=pooling)

# task-realted
fc1 = mx.symbol.FullyConnected(data= flat, num_hidden=2)  # dog or cat
softmax_activation1 = mx.symbol.SoftmaxActivation(data=fc1)
loss_1 = mx.symbol.softmax_cross_entropy(lhs=softmax_activation, rhs=label1)

concat = mx.symbol.Concat(flat, softmax_activation1, dim=1)  # concat info
fc2 = mx.symbol.FullyConnected(data= concat, num_hidden=10)  # dogs and cats, suppose 10 classes
loss_2 = mx.symbol.SoftmaxOutput(data=fc2, label=label2)
return mx.symbol.Group([loss1, loss2])
```

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
